src/core/border_router/routing_manager.cpp

```diff
@@ -373,22 +471,50 @@ bool RoutingManager::ContainsPrefix(const Ip6::Prefix &aPrefix, const Ip6::Prefi
     return ret;
 }
 
-const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void) const
+const Ip6::Prefix *RoutingManager::EvaluateOnLinkPrefix(void)
 {
-    const Ip6::Prefix *newOnLinkPrefix = nullptr;
+    const Ip6::Prefix *newOnLinkPrefix      = nullptr;
+    const Ip6::Prefix *smallestOnLinkPrefix = nullptr;
+    uint8_t            onLinkPrefixNum      = 0;
 
-    // We don't evaluate on-link prefix if we are doing
-    // Router Discovery or we have already discovered some
-    // on-link prefixes.
+    // We don't evaluate on-link prefix if we are doing Router Solicitation.
     VerifyOrExit(!mRouterSolicitTimer.IsRunning());
-    if (IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix))
+
+    for (uint8_t i = 0; i < mDiscoveredPrefixNum; ++i)
     {
-        otLogInfoBr("EvaluateOnLinkPrefix: there is already on-link prefix %s on interface %u",
-                    mDiscoveredOnLinkPrefix.ToString().AsCString(), mInfraIfIndex);
-        ExitNow();
+        if (!mDiscoveredPrefixes[i].mIsOnLinkPrefix)
+        {
+            continue;
+        }
+
+        ++onLinkPrefixNum;
+
+        if (smallestOnLinkPrefix == nullptr ||
+            IsPrefixSmallerThan(mDiscoveredPrefixes[i].mPrefix, *smallestOnLinkPrefix))
+        {
+            smallestOnLinkPrefix = &mDiscoveredPrefixes[i].mPrefix;
+        }
+    }
+
+    // We start advertising our local on-link prefix if there is no existing one.
+    if (onLinkPrefixNum == 0)
+    {
+        if (AddExternalRoute(mLocalOnLinkPrefix, OT_ROUTE_PREFERENCE_MED) == OT_ERROR_NONE)
+        {
+            newOnLinkPrefix = &mLocalOnLinkPrefix;
+        }
     }
+    // When an application-specific on-link prefix is received and it is bigger than the
+    // advertised prefix, we will not remove the advertised prefix. In this case, there
+    // will be two on-link prefixes on the infra link. But all BRs will still converge to
+    // the same smallest on-link prefix and the application-specific prefix is not used.
+    else if (mAdvertisedOnLinkPrefix != nullptr && *smallestOnLinkPrefix != *mAdvertisedOnLinkPrefix)
+    {
+        otLogInfoBr("EvaluateOnLinkPrefix: there is already smaller on-link prefix %s on interface %u",
+                    smallestOnLinkPrefix->ToString().AsCString(), mInfraIfIndex);
 
-    newOnLinkPrefix = &mLocalOnLinkPrefix;
+        RemoveExternalRoute(*mAdvertisedOnLinkPrefix);
```

Comment:
Question about this case, 
- we were advertising `mAdvertisedOnLinkPrefix` before in our RA.
- We now see there is a smaller prefix, so we try to remove ours
   - I don't think we send an RA with lifetime zero to remove ours  and basically expect the timeout to invalidate it
   - But we immediately remove the external route in thread net-data.
- So we can have a trasnistent time where the on-link prefix is still being used in infra net side (before timing out) and so devices on infra side can still use IP addresses based on this prefix to interact with Thread nodes, but this BR will immediately stop routing for this prefix (stop acting as external route for this prefix).
- Can this be a problem?

Reply:
I agree this can be an issue. We should continue to advertise external route for any prefixes that may still be in use.

Reply:
> I don't think we send an RA with lifetime zero to remove ours and basically expect the timeout to invalidate it

The `newOnLinkPrefix` is default to `nullptr` and it will be used for `SendRouterAdvertisement()` to send RA with zero lifetime. I may be more readable to have `SendRouterAdvertisement` together with `RemoveExternalRoute` but we didn't do it so that we can send all RA options in a single RA message.

But the issue still exists, per https://tools.ietf.org/html/rfc4862#section-5.5.3:
>1.  If the received Valid Lifetime is greater than 2 hours or
       greater than RemainingLifetime, set the valid lifetime of the
       corresponding address to the advertised Valid Lifetime.
>2.  If RemainingLifetime is less than or equal to 2 hours, ignore
       the Prefix Information option with regards to the valid
       lifetime, unless the Router Advertisement from which this
       option was obtained has been authenticated (e.g., via Secure
       Neighbor Discovery [RFC3971]).  If the Router Advertisement
       was authenticated, the valid lifetime of the corresponding
       address should be set to the Valid Lifetime in the received
       option.
>
>3.  Otherwise, reset the valid lifetime of the corresponding
          address to 2 hours.

SLAAC address will be retained even when RA with zero valid lifetime is received (zero valid lifetime is useful to tell other BRs to start advertising their own prefixes). My idea is to apply the same rule of SLAAC to external routes and delay removing external routes before the SLAAC addresses are actually expired. Thoughts?

Reply:
Please notice that even a SLAAC address is expired/deprecated, it doesn't stops the application layer to use it:
>Note that the feasibility of initiating new communication using a
   non-deprecated address may be an application-specific decision, as
   only the application may have knowledge about whether the (now)
   deprecated address was (or still is) in use by the application.  For
example, if an application explicitly specifies that the protocol
   stack use a deprecated address as a source address, the protocol
   stack must accept that; the application might request it because that
   IP address is used in higher-level communication and there might be a
   requirement that the multiple connections in such a grouping use the
   same pair of IP addresses.

https://tools.ietf.org/html/rfc4862#section-5.5.4

If we remove the external route for the on-link prefix when those SLAAC addresses on infra link are deprecated, then we will not be able to support the cases mentioned above (though I don't think it is a usual case).

Properly handling the cases requires non-trivial changes to current implementation. I am planning to leave this to future enhancement PR, will it work for you? @abtink @jwhui

Reply:
Added `TODO` for this issue.

Reply:
SGTM. Thanks.